src/TimeControl.cpp

```diff
@@ -49,24 +52,91 @@ TimeControl::TimeControl(int maintime, int byotime,
     reset_clocks();
 }
 
+std::string TimeControl::stones_left_to_text_sgf(const int color) const {
+    auto s = std::string{};
+    // We must be in byo-yomi before interpreting stones.
+    if (m_inbyo[color]) {
+        const auto c = color == FastBoard::BLACK ? "OB[" : "OW[";
+        if (m_byostones) {
+            s += c + std::to_string(m_stones_left[color]) + "]";
+        } else if (m_byoperiods) {
+            // KGS extension.
+            s += c + std::to_string(m_periods_left[color]) + "]";
+        }
+    }
+    return s;
+}
+
 std::string TimeControl::to_text_sgf() const {
     if (m_byotime != 0 && m_byostones == 0 && m_byoperiods == 0) {
-        return ""; // infinite
+        return ""; // Infinite time.
     }
-    auto s = "TM[" + std::to_string(m_maintime/100) + "]";
+    auto s = "TM[" + std::to_string(m_maintime / 100) + "]";
     if (m_byotime) {
         if (m_byostones) {
             s += "OT[" + std::to_string(m_byostones) + "/";
-            s += std::to_string(m_byotime/100) + " Canadian]";
+            s += std::to_string(m_byotime / 100) + " Canadian]";
         } else {
             assert(m_byoperiods);
             s += "OT[" + std::to_string(m_byoperiods) + "x";
-            s += std::to_string(m_byotime/100) + " byo-yomi]";
+            s += std::to_string(m_byotime / 100) + " byo-yomi]";
         }
+        s += stones_left_to_text_sgf(FastBoard::BLACK);
+        s += stones_left_to_text_sgf(FastBoard::WHITE);
     }
+    // Generously round up to avoid a remaining time of 0 triggering byo-yomi
+    // to be started when the sgf is loaded. This happens because byo-yomi
+    // stones have to be only written to the sgf when actually in byo-yomi
+    // and this is interpreted in adjust_time() as a special case
+    // that starts byo-yomi.
+    const auto black_time_left = (m_remaining_time[FastBoard::BLACK] + 99) / 100;
+    const auto white_time_left = (m_remaining_time[FastBoard::WHITE] + 99) / 100;
+    s += "BL[" + std::to_string(black_time_left) + "]";
+    s += "WL[" + std::to_string(white_time_left) + "]";
     return s;
 }
 
+std::shared_ptr<TimeControl> TimeControl::make_from_text_sgf(
+    const std::string& maintime, const std::string& byoyomi,
+    const std::string& black_time_left, const std::string& white_time_left,
+    const std::string& black_moves_left, const std::string& white_moves_left) {
+    const auto maintime_centis = std::stoi(maintime) * 100;
+    auto byotime = 0;
+    auto byostones = 0;
```

Comment:
Note that defaulting `byostones` to 0 here matches the behaviour of the GTP time settings handling but is inconsistent with the `TimeControl` constructor's default argument of 25. Why is it 25?

Reply:
It's what Leela assumes the timecontrol is if you don't configure anything.

Reply:
Sure but is there a reason why 25? I'm trying to understand if it matters whether I use 0 here or if it should also be 25. As far as I can tell, it doesn't matter what the value is because there are no byo-yomi periods if nothing is configured or e.g. the byo-yomi syntax isn't recognised.
So I went for 0 to match the behaviour when time settings are given by a GTP command.

Reply:
>Sure but is there a reason why 25?

That's a convenient thinking time for playing games against the computer if you have a primitive interface that can't even do a time_settings.

Reply:
@Hersmunch I guess we should also verify that if you load an SGF without time_settings (because the saver didn't know about them), the engine has a "reasonable" default.

Reply:
> > Sure but is there a reason why 25
>
> That's a convenient thinking time for playing games against the computer if you have a primitive interface that can't and won't even do a time_settings.

Time? Here's the time control constructor default arguments:
https://github.com/leela-zero/leela-zero/blob/63af3687462ba27daeabc526da6bbfa3d6149389/src/TimeControl.h#L29-L34
With no time or periods does it matter how many stones there are in byo-yomi?
After launching leelaz, if a `time_left b 1 0` is sent (1 sec, 0 stones == not in byo-yomi) and then the time is run down (e.g. via `auto`) the result is that byo-yomi is entered with 25 stones but then the clock continues to count down, going negative. This resets to 0 when the stones hit 0 and then goes negative again.

> So basically test that if you just launch the program or reset_board or change board size etc and don't do any time_settings, you still get something sane.

These all result in the sgf containing `TM[3600]BL[3600]WL[3600]`. Notice that this doesn't include any byo-yomi. This means that if the sgf is loaded again that it doesn't have the behaviour described above and instead doesn't go into byo-yomi (as per `byostones = 0` in my code) but the clock goes negative from then on.
I think it would be good to make this behaviour consistent with the time settings used just after launching leelaz but perhaps that should be changed to have some time?

> @Hersmunch I guess we should also verify that if you load an SGF without time_settings (because the saver didn't know about them), the engine has a "reasonable" default.

This didn't behave as I expected, which was that no time settings would be loaded and the time settings would be whatever leelaz was using before loading the sgf. Instead, the time settings are reset to the default behaviour after launching leelaz. Am I right in thinking that this is because the shared_ptr will point to a time control based upon the default constructor?

Reply:
Okay, probably byostones should just be 0 by default given that there was no time set.

Reply:
>Am I right in thinking that this is because the shared_ptr will point to a time control based upon the default constructor?

No, it will just be null.

What happens is that this case is detected and the set_timecontrol isn't done here: https://github.com/leela-zero/leela-zero/pull/2172/files#diff-44ba5da097788068a1c8def978c3293cR84

And you indeed get the default constructor.